dispatcher/service/next_wakeup_runner.py

```diff
@@ -0,0 +1,123 @@
+            method_name = getattr(process_object, '__name__', str(process_object))
+            self.name = f'next-run-manager-of-{method_name}'
+        else:
+            self.name = name
+
+    async def process_wakeups(self, current_time: float, do_processing: bool = True) -> Optional[float]:
+        """Runs process_object for objects past for which we have passed the wakeup time
+
+        Returns the time of the soonest wakeup that has not been processed here
+
+        Arguments:
+         - current_time - output of time.monotonic() passed from caller to keep this deterministic
+         - do_processing - to help optimizations, False value allows checking next wakeup time
+           without calling the callback for anything
+        """
+        future_wakeup = None
+        for obj in self.wakeup_objects:
+            if obj_wakeup := obj.next_wakeup():
+                if do_processing and (obj_wakeup < current_time):
+                    await self.process_object(obj)
+                    # refresh wakeup, which should be nullified or pushed back by process_object
+                    obj_wakeup = obj.next_wakeup()
+                    if obj_wakeup is None:
+                        continue
+                if (future_wakeup is None) or obj_wakeup < future_wakeup:
+                    future_wakeup = obj_wakeup
+        return future_wakeup
+
+    async def background_task(self) -> None:
+        while not self.shutting_down:
+            now_time = time.monotonic()
+            next_wakeup = await self.process_wakeups(now_time)
+            if next_wakeup is None:
+                return
+
+            delta = next_wakeup - now_time
+            if delta <= 0.0:
+                logger.info(f'Wakeup processor {self.name} has leftover wakeups, {delta}s in past, sleeping for 0.1s')
+                delta = 0.1
+
+            try:
+                await asyncio.wait_for(self.kick_event.wait(), timeout=delta)
+            except asyncio.TimeoutError:
+                pass  # intended mechanism to hit the next schedule
+            except asyncio.CancelledError:
+                logger.info(f'Task {self.name} cancelled, returning')
+                return
+
+            self.kick_event.clear()
+
+    def mk_new_task(self) -> None:
+        """Should only be called if a task is not currently running"""
+        self.asyncio_task = asyncio.create_task(self.background_task(), name=self.name)
+        ensure_fatal(self.asyncio_task)
+
+    def kick(self) -> None:
+        """Initiates the asyncio task to wake up at the next run time
+
+        This needs to be called if objects in wakeup_objects are changed, for example
+        """
+        if self.process_wakeups(current_time=time.monotonic(), do_processing=False) is None:
+            # Optimization here, if there is no next time, do not bother managing tasks
+            return
+        if self.asyncio_task:
+            if self.asyncio_task.done():
+                self.mk_new_task()
```

Comment:
is this to allow restarting the background task after it has been cancelled?

Reply:
Not canceling, from the `return`. When the task does a `return` then it still exists as task. `.done()` will return `True`, and `.result()` will give the return value (which we don't care about). If there are no future wakeups, then the task does a `return`, so it's still hanging out but it's left. This is done specifically thinking about "delay" and "timeout" features which will only be used for a specific task once-in-a-while, and for some apps, never used at all. So I'd rather not have an active task in those cases.